webrtc/README.md

```diff
@@ -0,0 +1,445 @@
+   compatibility can be found
+   [here](https://developer.mozilla.org/en-US/docs/Web/API/RTCCertificate). In
+   practice, this is not an issue since the fingerprint is embedded in the local
+   SDP string.
+
+3. On success of the authentication handshake, the used datachannel is
+   closed and the plain WebRTC connection is used with its multiplexing
+   capabilities via datachannels. See [Multiplexing](#multiplexing).
+
+Note: WebRTC supports different hash functions to hash the TLS certificate (see
+https://datatracker.ietf.org/doc/html/rfc8122#section-5). The hash function used
+in WebRTC and the hash function used in the multiaddr `/certhash` component MUST
+be the same. On mismatch the final Noise handshake MUST fail.
+
+_A_ knows _B_'s fingerprint hash algorithm through _B_'s multiaddr. _B_ could
+infer _A_'s fingerprint hash algorithm through _A_'s TLS certificate signature.
+Or _B_ could assume _A_ to use the same hash algorithm it discovers through
+_B_'s multiaddr. For now implementations MAY assume both _A_ and _B_ to use
+sha-256. Future iterations of this specification may add support for other hash
+algorithms in a backwards compatible way.
+
+Implementations SHOULD setup all the necessary callbacks (e.g.
+[`ondatachannel`](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/datachannel_event))
+before starting the Noise handshake. This is to avoid scenarios like one where
+_A_ initiates a stream before _B_ got a chance to set the `ondatachannel`
+callback. This would result in _B_ ignoring all the messages coming from _A_
+targeting that stream.
+
+Implementations MAY open streams before completion of the Noise handshake.
+Applications MUST take special care what application data they send, since at
+this point the peer is not yet authenticated. Similarly, the receiving side MAY
+accept streams before completion of the handshake.
+
+## Previous, ongoing and related work
+
+- Work in progress implementations of this specification:
+  - https://github.com/little-bear-labs/js-libp2p-webrtc/
+  - https://github.com/libp2p/go-libp2p/pull/1655
+  - https://github.com/libp2p/rust-libp2p/pull/2622
+- Past related work:
+  - Proof of concept for the server side (native) and the client side (Rust in
+    WASM): https://github.com/wngr/libp2p-webrtc
+  - WebRTC using STUN and TURN: https://github.com/libp2p/js-libp2p-webrtc-star
+
+## Test vectors
+
+### Noise prologue
+
+All of these test vectors represent hex-encoded bytes.
+
+#### 1. Both client and server use SHA-256
+
+```
+client_fingerprint = "3e79af40d6059617a0d83b83a52ce73b0c1f37a72c6043ad2969e2351bdca870"
+server_fingerprint = "30fc9f469c207419dfdd0aab5f27a86c973c94e40548db9375cca2e915973b99"
+
+prologue = "6c69627032702d7765627274632d6e6f6973653a12203e79af40d6059617a0d83b83a52ce73b0c1f37a72c6043ad2969e2351bdca870122030fc9f469c207419dfdd0aab5f27a86c973c94e40548db9375cca2e915973b99"
+```
```

Comment:
@ckousik @John-LittleBearLabs might be helpful to add as unit tests to the Golang and JS implementation. Thanks @thomaseizinger for adding it.

Reply:
Perhaps we should add here that in relation to the spec, `A == client` and `B == server`.